service/s3/api_client.go

```diff
@@ -150,18 +150,22 @@ func NewFromConfig(cfg aws.Config, optFns ...func(*Options)) *Client {
 	return New(opts, optFns...)
 }
 
-func resolveHTTPClient(o *Options) {
-	if o.HTTPClient != nil {
+func resolveRetryer(o *Options) {
+	if o.Retryer != nil {
 		return
 	}
-	o.HTTPClient = aws.NewBuildableHTTPClient()
+	o.Retryer = retry.NewStandard()
 }
 
-func resolveRetryer(o *Options) {
-	if o.Retryer != nil {
+func resolveHTTPClient(o *Options) {
+	if o.HTTPClient != nil {
 		return
 	}
-	o.Retryer = retry.NewStandard()
+	client := aws.NewBuildableHTTPClient()
+	// set client transport's DisableCompression to true. This will instruct the HTTP
+	// Client to not auto decode the GZIP response.
+	client.GetTransport().DisableCompression = true
+	o.HTTPClient = client
 }
```

Comment:
We'll want to document this behavior. What do you think about using a pattern similar to DynamoDB's disable compression customization? If the customer were to create a custom HTTP client or a custom transport this method would not work, and the behavior of the S3 API client would change significantly, in a non-obvious way.  This was the reason the DDB's customization set the `Accept-Encoding` header.

Its not 100% clear if s3 should have the option to enable compression or not though. future behavior of the s3 client like trailing checksum validation might get overly complicated if the API client support enabling decompressing responses.

Reply:
Moved the dynamodb middleware to service/internal/customizations so as to reuse it ; For s3 we won't generate Enable option and default to disabling